Libraries/pbxbuild/Sources/Phase/ModuleMapResolver.cpp

```diff
@@ -29,143 +30,64 @@ Phase::ModuleMapResolver::
-{
-    std::string inputModuleMapPath = environment.resolve("MODULEMAP_FILE");
-
-    if (!inputModuleMapPath.empty()) {
-        /*
-         * Copy in the input module map as an auxiliary file.
-         */
-        std::string path = FSUtil::ResolveRelativePath(inputModuleMapPath, workingDirectory);
-        return Tool::Invocation::AuxiliaryFile::Chunk::File(path);
-    } else {
-        /*
-         * Find umbrella header: a header with the same name as the target's module.
-         */
-        std::string moduleName = environment.resolve("PRODUCT_MODULE_NAME");
-        ext::optional<std::string> umbrellaHeaderName = FindUmbrellaHeaderName(environment, moduleName, target);
-
-        if (umbrellaHeaderName) {
-            /*
-             * Generate a module map as an auxiliary file.
-             */
-            auto contents = std::string();
-            contents += "framework module " + moduleName + " {\n";
-            contents += "  umbrella header \"" + *umbrellaHeaderName + "\"\n";
-            contents += "  \n";
-            contents += "  export *\n";
-            contents += "  module * { export * }\n";
-            contents += "}\n";
-
-            auto data = std::vector<uint8_t>(contents.begin(), contents.end());
-            return Tool::Invocation::AuxiliaryFile::Chunk::Data(data);
-        } else {
-            return ext::nullopt;
-        }
-    }
+    /* Copy in the module map as part of the build. */
+    copyResolver->resolve(toolContext, environment, { entry.intermediatePath() }, FSUtil::GetDirectoryName(entry.finalPath()), "Ditto");
 }
 
 bool Phase::ModuleMapResolver::
 resolve(Phase::Environment const &phaseEnvironment, Phase::Context *phaseContext) const
 {
     pbxsetting::Environment const &environment = phaseEnvironment.targetEnvironment().environment();
-    std::string const &workingDirectory = phaseContext->toolContext().workingDirectory();
 
-    /*
-     * Check if creating module maps is even requested.
-     */
-    if (!pbxsetting::Type::ParseBoolean(environment.resolve("DEFINES_MODULE"))) {
-        return true;
+    std::unique_ptr<Tool::ModuleMapResolver> moduleMapResolver = Tool::ModuleMapResolver::Create();
+    if (moduleMapResolver == nullptr) {
+        return false;
     }
 
-    std::string intermediateModuleMapRoot = environment.resolve("TARGET_TEMP_DIR");
-    // TODO: what about non-framework product types?
-    std::string finalModuleMapRoot = environment.resolve("TARGET_BUILD_DIR") + "/" + environment.resolve("CONTENTS_FOLDER_PATH") + "/" + "Modules";
+    Tool::CopyResolver const *copyResolver = phaseContext->copyResolver(phaseEnvironment);
+    if (copyResolver == nullptr) {
+        fprintf(stderr, "warning: failed to get copy tool for module map\n");
+        return false;
+    }
 
     /*
-     * Handle the standard module map, either copied or generated.
+     * Create the module map info. Note this is required even if a module is not defined, in order to
```

Comment:
This reads as "a module map is needed for builds that don't define a module."  Seems weird.